File: aoc-python/day04/part1.py
from __future__ import annotations

import argparse
import os.path

import pytest

import support

INPUT_TXT = os.path.join(os.path.dirname(__file__), "input.txt")


def compute(s: str) -> int:
    lines = s.splitlines()
    count = 0
    for line in lines:
        elves = [e.split("-") for e in line.split(",")]
        elves = sorted((list(map(int, e)) for e in elves), key=lambda x: x[0])
        (a, b), (c, d) = elves
        if (
            c <= d <= b
            or c <= a <= b <=  d
        ):
            count += 1


    return count

     
INPUT_S = """\
2-4,6-8
2-3,4-5
5-7,7-9
2-8,3-7
6-6,4-6
2-6,4-8
95-96,92-100
92-100,95-96
"""
EXPECTED = 4


@pytest.mark.parametrize(
    ("input_s", "expected"),
    ((INPUT_S, EXPECTED),),
)
def test(input_s: str, expected: int) -> None:
    assert compute(input_s) == expected


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("data_file", nargs="?", default=INPUT_TXT)
    args = parser.parse_args()

    with open(args.data_file) as f, support.timing():
        print(compute(f.read()))

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: aocJulia/days/day04/part1.jl
using Pipe, Test

using aocJulia: helpers


function compute(s)
    lines = @pipe s |> split |> split.(_, ",") |> map.(x->split(x, "-"), _)
    n = 0
    for pair in lines
        ((a, b), (c, d)) = (parse.(Int, e) for e in pair)
        if (
            a<=c<=d<=b
            || c<=a<=b<=d
        )
            n +=1 
        end
    end
    return n
end


INPUTS_S = """
2-4,6-8
2-3,4-5
5-7,7-9
2-8,3-7
6-6,4-6
2-6,4-8
"""
EXPECTED = 2

# @test compute(INPUTS_S) == EXPECTED

INPUT_TXT = helpers.read_input(dirname(Base.source_path()))
println(compute(INPUT_TXT))